xarray/core/indexing.py

```diff
@@ -434,7 +492,84 @@ def __setitem__(self, key, value):
         self.array[key] = value
 
 
-def orthogonally_indexable(array):
+def unbroadcast_indexes(key, shape):
+    """
+    Convert broadcasted indexers to orthogonal indexers.
+    If there is no valid mapping, raises IndexError.
+
+    key is usually generated by Variable._broadcast_indexes.
+
+    key: tuple of np.ndarray, slice, integer
+    shape: shape of array
+    """
+    key = expanded_indexer(key, len(shape))
+
+    if all(isinstance(k, integer_types + (slice,)) for k in key):
+        # basic indexing
+        return key
+
+    # Here, we need to distinguish integer and 1-element array
```

Comment:
> I guess more backends would support the fancy indexing in the future?

Maybe? It seems like we have at least three likely cases:
1. Backends support only "basic indexing "(int and slice). This is pretty common.
2. Backends support some of the "advanced indexing" use cases but not everything (e.g., restricted to most one list). This is also pretty common (e.g., dask and h5py).
3. Backends support "orthogonal indexing" instead of advanced indexing. NetCDF4 does this (but perform can be pretty terrible).
4. Backends support NumPy's fully vectorized "advanced indexing". This is quite rare -- I've only seen this for backends that actually store their data in the form of NumPy arrays (e.g., `scipy.io.netcdf`).